src/core/mac/mac.cpp

```diff
@@ -2025,5 +2145,225 @@ uint8_t Mac::GetTimeIeOffset(const Frame &aFrame)
+{
+#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+    return mSubMac.IsCslStarted() && !IsCslIeSuppressed() &&
+           !Get<Mle::Mle>().IsRxOnWhenIdle(); // Only for certification
+#else
+    return mSubMac.IsCslStarted() && !Get<Mle::Mle>().IsRxOnWhenIdle();
+#endif // OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+}
+
+otError Mac::StartCsl(void)
+{
+    otError   error  = OT_ERROR_NONE;
+    Neighbor *parent = &Get<Mle::Mle>().GetParent();
+
+    VerifyOrExit(!mRxOnWhenIdle && Get<Mle::MleRouter>().GetRole() == OT_DEVICE_ROLE_CHILD,
+                 error = OT_ERROR_INVALID_STATE);
+    VerifyOrExit(parent->IsSupportEnhancedFramePending(), error = OT_ERROR_NOT_CAPABLE);
+
+    otPlatRadioEnableCsl(&GetInstance(), mSubMac.GetCslPeriod(), &parent->GetExtAddress());
+
+    SuccessOrExit(error = mSubMac.StartCsl());
+
+    Get<Mle::Mle>().ScheduleChildUpdateRequest();
+
+exit:
+    return error;
+}
+
+void Mac::StopCsl(void)
+{
+    Neighbor *parent = &Get<Mle::Mle>().GetParent();
+    VerifyOrExit(parent != NULL);
+
+    VerifyOrExit(mSubMac.IsCslStarted()); // This is required to avoid StopCsl is called during initialization and call
+                                          // otPlatRadioEnableCsl
+
+    mSubMac.StopCsl();
+
+    otPlatRadioEnableCsl(&GetInstance(), 0, &parent->GetExtAddress());
+
+exit:
+    return;
+}
+#endif // OPENTHREAD_CONFIG_CSL_RECEIVER_ENABLE
+
+#if OPENTHREAD_CONFIG_CSL_TRANSMITTER_ENABLE
+void Mac::ProcessCsl(const RxFrame &aFrame, const Address &aSrcAddr)
+{
+    const uint8_t *cur   = aFrame.GetHeaderIe(Frame::kHeaderIeCsl);
+    Child *        child = Get<ChildTable>().FindChild(aSrcAddr, Child::kInStateAnyExceptInvalid);
+    const CslIe *  csl   = reinterpret_cast<const CslIe *>(cur + sizeof(HeaderIe));
+
+    VerifyOrExit(cur != NULL && child != NULL && aFrame.GetSecurityEnabled());
+
+    child->SetCslPeriod(csl->GetPeriod());
+    // Use ceiling to ensure the the time diff will be within OT_US_PER_TEN_SYMBOLS
+    child->SetCslPhase(((aFrame.GetTimestamp() + OT_US_PER_TEN_SYMBOLS - 1) / OT_US_PER_TEN_SYMBOLS + csl->GetPhase()) %
+                       csl->GetPeriod());
+    child->SetCslSynchronized(true);
+    child->SetCslLastHeard(TimerMilli::GetNow());
+    otLogDebgMac("Timestamp=%lu Sequence=%u CslPeriod=%hu CslPhase=%hu TransmitPhase=%hu",
+                 static_cast<uint32_t>(aFrame.GetTimestamp()), aFrame.GetSequence(), csl->GetPeriod(), csl->GetPhase(),
+                 child->GetCslPhase());
+
+exit:
+    return;
+}
+#endif // OPENTHREAD_CONFIG_CSL_TRANSMITTER_ENABLE
+
+#if OPENTHREAD_CONFIG_MAC_HEADER_IE_SUPPORT
+otError Mac::AppendHeaderIe(TxFrame &aFrame, bool aIsTimeSync) const
+{
+    OT_UNUSED_VARIABLE(aFrame);
+    OT_UNUSED_VARIABLE(aIsTimeSync);
+
+    const size_t kMaxNumHeaderIe = 3; // TimeSync + Csl + Termination2
+    HeaderIe     ieList[kMaxNumHeaderIe];
+    otError      error   = OT_ERROR_NONE;
+    uint8_t      ieCount = 0;
+
+    VerifyOrExit(aFrame.IsIePresent());
+
```

Comment:
Is it necessary to check FCF to make sure it is version 0b10 or later in case a ver 0b00/0b01 frame is passed in?